src/V3File.cpp

```diff
@@ -916,19 +917,142 @@ void V3OutFormatter::printf(const char* fmt...) {
+    while (aio_suspend(aiocb_list, nitems, nullptr) < 0) {
+        if (errno != EINTR) v3fatal("Asynchronous write failure: " << std::strerror(errno));
+    }
+}
+
+static void waitForCompletionInternal(aiocb** startp, aiocb** endp) {
+    while (startp != endp) {
+        // Wait for any queued buffer to complete, ignore wake-ups caused by signals
+        aioSuspendIgnoreInterrupts(startp, std::distance(startp, endp));
+        // Find completed operations
+        for (aiocb** p = startp; p != endp; ++p) {
+            if (aioOperationHasCompleted(**p)) {
+                *p = *startp;
+                ++startp;
+            }
+        }
+    }
+}
+
+AioWriteBufferHandle AioWriter::dequeueBuffer() {
+    // No free buffers, but total under limit - allocate and return new buffer.
+    if (s_freeBuffers.empty() && s_allocatedBuffersCount < BUFFERS_COUNT_MAX) {
+        s_buffersInUse.emplace_back();
+        ++s_allocatedBuffersCount;
+        return AioWriteBufferHandle(std::prev(s_buffersInUse.end()));
+    }
+
+    // There is a free buffer, return it.
+    if (!s_freeBuffers.empty()) {
+        auto iter = s_freeBuffers.begin();
+        s_buffersInUse.splice(s_buffersInUse.end(), s_freeBuffers, iter);
+        iter->clear();
+        return AioWriteBufferHandle(iter);
+    }
+
+    aiocb* aiocbpArray[s_buffersInKernel.size()];
+    auto queuedBufferIter = s_buffersInKernel.begin();
+    for (auto& cbp : aiocbpArray) {
+        cbp = &queuedBufferIter->m_aiocb;
+        ++queuedBufferIter;
+    }
+    // Wait for any queued buffer to complete.
+    aioSuspendIgnoreInterrupts(aiocbpArray, s_buffersInKernel.size());
+
+    // Find first buffer that completed.
+    auto iter = s_buffersInKernel.begin();
+    for (; iter != s_buffersInKernel.end(); ++iter) {
+        if (aioOperationHasCompleted(iter->m_aiocb)) break;
+    }
+    UASSERT((iter != s_buffersInKernel.end()), "aio_suspend returned without available buffer");
+    s_buffersInUse.splice(s_buffersInUse.end(), s_buffersInKernel, iter);
+    iter->clear();
+    return AioWriteBufferHandle(iter);
+}
+
+void AioWriter::queueBuffer(AioWriteBufferHandle handle) {
+    auto iter = handle.m_iter;
+    if (iter->size() == 0) {
+        s_freeBuffers.splice(s_freeBuffers.end(), s_buffersInUse, iter);
+        return;
+    }
+    s_buffersInKernel.splice(s_buffersInKernel.end(), s_buffersInUse, iter);
+    UASSERT((iter->m_aiocb.aio_fildes >= 0), "Buffer queued with invalid FD");
+    if (aio_write(&iter->m_aiocb) < 0)
+        v3fatal("Asynchronous write failure: " << std::strerror(errno));
+}
+
+void AioWriter::waitForCompletion() {
+    aiocb* aiocbpArray[s_buffersInKernel.size()];
+    aiocb** startp = &aiocbpArray[0];
+    aiocb** endp = startp;
+
+    for (auto& buf : s_buffersInKernel) { *endp++ = &buf.m_aiocb; }
+
+    waitForCompletionInternal(startp, endp);
+    s_freeBuffers.splice(s_freeBuffers.end(), s_buffersInKernel);
+}
+
+void AioWriter::waitForCompletion(int fd) {
```

Comment:
A lot of duplicate code with above function, maybe combine and have bool as to if fd is provided?